samples/custom_dispatch/xnnpack/plugin/system_plugin.c

```diff
@@ -44,8 +63,82 @@ typedef struct {
   iree_hal_executable_plugin_allocator_t host_allocator;
   FILE* file;
   pthreadpool_t threadpool;
+  struct cache* cache;
 } system_plugin_t;
 
+inline static void log_debug(const char* format, ...) {
+#ifdef XNNPACK_PLUGIN_LOG
+  va_list args;
+  va_start(args, format);
+  vfprintf(stderr, format, args);
+  fprintf(stderr, "\n");
+  va_end(args);
+#endif
+}
+
+static struct cache* create_cache(size_t max_size) {
+  struct cache* m = malloc(sizeof(struct cache));
+  m->max_size = max_size;
+  m->size = 0;
+  m->keys = malloc(sizeof(struct cache_key) * max_size);
+  m->values = malloc(sizeof(struct cache_value) * max_size);
+  pthread_mutex_init(&m->lock, NULL);
+  return m;
+}
+
+static void destroy_cache(struct cache* m) {
+  log_debug("Max cache size used: %zu", m->size);
+  for (size_t i = 0; i < m->size; i++) {
+    xnn_delete_operator(m->values[i].op);
+    free(m->values[i].kernel_scale);
+  }
+  pthread_mutex_destroy(&m->lock);
+  free(m->keys);
+  free(m->values);
+  free(m);
+}
+
+inline static void cache_dump_key(const struct cache_key* key) {
+  log_debug("cache_key { kernel_id=%zu }", key->kernel_id);
+}
+
+static bool cache_insert(struct cache* m, struct cache_key key,
+                         struct cache_value value) {
+  if (m->size >= m->max_size) {
+    log_debug("unimplemented: variable size cache");
```

Comment:
Add a TODO comment.

I see your pain, one way is to use a c++ and expose the std::unordered_map with C APIs.

You may use a linked list since the lookup is already using a linear traversal below. (It is okay for now.)

(But there might be some restriction in the runtime development to be portable without using a C++ library.)

Reply:
Yeah, I'm also happy to implement variable sizing using the standard "double the storage array every time you run out of space". I was just trying to get things going with this PR, and optimize later.